Parse, algebrize, and translate basic numeric predicate patterns

rnewman opened the pull request:
This PR starts to split up `cc.rs` a little, as well as doing some cleanup. (I left in some reformatting that VSCode did on its own, mostly in their own commits.)

I then added a test for some things that CC isn't smart enough to do yet (see #373), and finally moved on to the meat of the PR: extending the parser to handle predicates.

Right now `edn::Value::List` isn't 'spreadable' into a slice, which our parsing approach relies on. That means we don't yet support the syntax you usually see for predicates:

```
[(< ?x 10)]
```

Instead we partly follow the Datomic query grammar (!!):

```
[] = list or vector
…
pred-expr = [ [pred fn-arg+] ]
```

and support only vectors: `[[< ?x 10]]` is valid, and that's what we handle. I'll expand that next.

ncalexan commented:
> Right now edn::Value::List isn't 'spreadable' into a slice, which our parsing approach relies on. That means we don't yet support the syntax you usually see for predicates:

This might be an argument for backing `edn::Value::List` with a contiguous data structure.  It would not be hard to `impl Parser<T> for LinkedList<T>`, but I doubt it's worth it.

ncalexan commented on query-algebrizer/src/cc.rs:
```diff
@@ -1235,4 +1235,114 @@ mod testing {
     }
+
+    #[test]
+    fn test_fulltext_type_disagreement() {
```
Let's start `/// Noting what the test verifies`.  I'm guilty of not doing this as well.

ncalexan commented on query-algebrizer/src/cc.rs:
```diff
@@ -1235,4 +1235,114 @@ mod testing {
+
+        assert!(cc.is_known_empty);
+        assert_eq!(cc.empty_because.unwrap(),
+                   EmptyBecause::TypeMismatch(y.clone(), ValueType::String, ValueType::Boolean));
```
Nice, you made this more structured.

ncalexan commented on query-algebrizer/src/cc.rs:
```diff
@@ -1235,4 +1235,114 @@ mod testing {
+    }
+
+    #[test]
+    #[should_panic(expected = "assertion failed: cc.is_known_empty")]
```
Is this better than explicitly asserting the current state and explaining what the future state should be?  It definitely doesn't allow you to add to the test easily.

ncalexan commented on query-parser/src/parse.rs:
```diff
@@ -136,21 +178,25 @@ def_value_parser_fn!(Where, pattern, Pattern, input, {
-            patterns.into_iter().map(WhereClause::Pattern).collect()
-        })
+    // Right now we only support patterns and predicates. See #239 for more.
+    (many1::<Vec<WhereClause>, _>(choice([Where::pattern(), Where::pred()])), eof())
```
There's a weirdness using `combine` where `choice` and parsers that consume some input  require care.  Have you taken that care here?  (I don't know the details offhand so I can't determine if you have without doing some research myself, so I'd like to save that research if you have done it.)

ncalexan commented on query/src/lib.rs:
```diff
@@ -129,6 +150,19 @@ pub enum FnArg {
+impl FromValue<FnArg> for FnArg {
+    fn from_value(v: &edn::Value) -> Option<FnArg> {
+        // TODO: support SrcVars.
+        Variable::from_value(v)
```
Is this better than matching the `edn::Value` first?  You don't want to encode knowledge of variables into `FnArg`?

ncalexan commented on query/src/lib.rs:
```diff
@@ -433,6 +467,11 @@ impl Pattern {
 
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Predicate {
+    pub operator: PlainSymbol,
```
You're purposely accepting arbitrary predicate operators, right?  You'll constrain that set after parsing?  (Perhaps during interpretation/simplification?)

ncalexan commented on query-parser/src/parse.rs:
```diff
@@ -93,34 +96,73 @@ impl<I> Query<I>
+    match x {
+        edn::Value::Vector(mut v) => {
+            match v.pop() {
+                Some(edn::Value::Vector(items)) => Some(items),
```
You're returning consuming your inputs and returning owned values; can't you match `Some(edn::Value::List(linked_list))` and produce a `Vec` out of it?

ncalexan commented on query-parser/src/find.rs:
```diff
@@ -201,25 +201,21 @@ mod test_parse {
         let truncated_input = edn::Value::Vector(vec![Value::from_keyword(None, "find")]);
         assert!(parse_find(truncated_input).is_err());
 
-        let input = edn::Value::Vector(vec![
```
I feel like our editors are going to do some battle royale at some point, but that's in the future.  Right?  Right!

ncalexan commented on query-algebrizer/src/errors.rs:
```diff
@@ -0,0 +1,40 @@
+            display("invalid number of arguments to {}: expected {}, got {}.", name, expected, number)
+        }
+
+        UnboundVariable(var: Variable) {
```
Can you label this with the function `name`?

ncalexan commented on query-algebrizer/src/cc.rs:
```diff
@@ -822,6 +789,64 @@ impl ConjoiningClauses {
+    /// - Ensures that the predicate functions name a known operator.
+    /// - Accumulates a `BinaryPredicate` constraint into the `wheres` list.
+    pub fn apply_predicate<'s, 'p>(&mut self, schema: &'s Schema, predicate: Predicate) -> Result<()> {
+        if predicate.args.len() != 2 {
```
For these finite-length destructurings, I quite like `match (it.next(), it.next(), it.next()) { (Some, Some, None) => happy; _ => sad }`.

ncalexan commented on query-algebrizer/src/types.rs:
```diff
@@ -85,4 +93,99 @@ impl QualifiedAlias {
+            },
+
+            &BinaryPredicate { operator, ref left, ref right } => {
+                write!(f, "{:?} {:?} {:?}", left, operator, right)
```
I would be slightly to keep `{:?}` "standard", and overload `{}`.

rnewman commented:
Reworked with the things I could fix, rebased on top of the `unique` changes, and extended to do soup-to-nuts. Take another look, @ncalexan and @fluffyemily!

rnewman commented:
Oh, and this handles lists now.

ncalexan commented on query-parser/src/parse.rs:
```diff
@@ -93,34 +96,75 @@ impl<I> Query<I>
+def_value_parser_fn!(Where, pred, WhereClause, input, {
+    satisfy_map(|x: edn::Value| {
+        // Accept either a list or a vector here:
+        // `[(foo ?x ?y)]` or `[[foo ?x ?y]]`
```
What's the motivation here?  Is the intention to agree with Datomic, which probably takes anything sequential?

rnewman replied:
Indeed, the grammar declares that anywhere you see `[]` it means a list or vector (probably because all Clojure cares about are `seq`s). The common usages we see of things like `[(< ?x 5)]` are just convention.

I don't think we should jump to being so flexible; I'm happy to pave the cowpaths and make `[]` required in most cases.

ncalexan commented on query-parser/src/parse.rs:
```diff
@@ -93,34 +96,75 @@ impl<I> Query<I>
+                        args: args,
+                    })
+            });
+            let r: ParseResult<WhereClause, _> = p.parse_lazy(&items[..]).into();
```
I wonder if we should add `result()` to `ParseResult` or something to shorthand this `.ok().map()` pattern.

rnewman replied:
`ParseResult` is just a type alias to `Result`, so we can't easily add traits to it in `combine`. We could perhaps macroize this in parser utils… but in the mean time I'll use the `to_parsed_value` function I defined to at least make it more readable.

ncalexan commented on query-algebrizer/src/cc.rs:
```diff
@@ -822,6 +789,80 @@ impl ConjoiningClauses {
+        }
+    }
+
+    /// There are several kinds of predicates/functions in our Datalog:
```
nit: Datalog dialect?  Datalog vocabulary?

ncalexan commented on query-algebrizer/src/cc.rs:
```diff
@@ -822,6 +789,80 @@ impl ConjoiningClauses {
+        // To do so needs a slightly more sophisticated representation of type constraints — a set,
+        // not a single `Option`.
+
+        // TODO: static evaluation.
```
I'm glad you're punting on this.  Maybe link to a ticket?

ncalexan commented on query-algebrizer/src/types.rs:
```diff
@@ -85,4 +90,117 @@ impl QualifiedAlias {
+            &LessThanOrEquals => "<=",
+            &GreaterThan => ">",
+            &GreaterThanOrEquals => ">=",
+            &NotEquals => "!=",                // Datalog uses !=. SQL uses <>.
```
FWIW, I believe SQLite accepts the non-SQL-standard `!=`.  I prefer to use the SQL standard `<>`, as you have done here.

rnewman replied:
It does; one of SQLite's design goals is to do the best it can with SQL written for other database systems. But they all handle `<>`, so that's what we'll use.

ncalexan commented on query-translator/tests/translate.rs:
```diff
@@ -178,3 +178,50 @@ fn test_unknown_ident() {
+        value_type: ValueType::Double,
+        ..Default::default()
+    });
+
```
nit: you have a few extra newlines, here and below.

ncalexan commented on query-algebrizer/src/lib.rs:
```diff
@@ -62,37 +68,42 @@ impl AlgebraicQuery {
-            unimplemented!();
+        match where_clause {
+            WhereClause::Pattern(p) => {
+                cc.apply_pattern(schema, p);
```
Consider making `apply_pattern -> Result<>` as well, just for uniformity.

ncalexan commented on query-algebrizer/src/cc.rs:
```diff
@@ -83,6 +87,12 @@ impl<T> OptionEffect<T> for Option<T> {
     }
 }
 
+fn unit_type_set(t: ValueType) -> HashSet<ValueType> {
```
Don't you wish Rust made this easier?  Like `set.with(...)` or `Set::of{1,2,3}(...)`.

ncalexan replied:
I'd take a trait in `util` for making some of these easier.  Could probably do this generically across collection types with a `FromIterator` bound.

rnewman replied:
You can do `vec![1, 2].into_iter().collect()` with the right type bounds, but it's messy.

ncalexan commented on query-algebrizer/src/cc.rs:
```diff
@@ -247,6 +261,16 @@ impl ConjoiningClauses {
     }
 
+    /// Return a single `ValueType` if the given variable is known to have a precise type.
+    /// Returns `None` if the type of the variable is known but not precise -- "double
```
nit: or if the type of the variable is not known.

ncalexan commented on query-algebrizer/src/cc.rs:
```diff
@@ -792,6 +856,38 @@ impl ConjoiningClauses {
+            Constant(NonIntegerConstant::Text(_)) |
+            Constant(NonIntegerConstant::BigInteger(_)) => {
+                self.mark_known_empty(EmptyBecause::NonNumericArgument);
+                // We use Double because… well, we only have one slot!
```
I don't follow this comment.

rnewman replied:
That's because I simplified the code! I originally generated an error that said, in short, "this thing isn't a double", when of course it really meant "this isn't a {double, long}".

Comment removed!

ncalexan commented on query-translator/tests/translate.rs:
```diff
@@ -186,9 +186,8 @@ fn test_numeric_less_than_unknown_attribute() {
-    // TODO: we don't infer numeric types from numeric predicates, because the _SQL_ type code
-    // is a single value (5), but the Datalog types are a set (Double and Long).
-    // When we do, this will correctly use `datoms` instead of `all_datoms`.
+    // Although we infer numericness from numeric predicates, we've already assigned a table to the
```
This seems like a TODO at least.

rnewman commented:
> `+    // Although we infer numericness from numeric predicates, we've already assigned a table to the`
>
> This seems like a TODO at least.

I drafted an issue for this while I was on the plane, but didn't file it yet.

I'm on the fence with this.

On the one hand, we can solve this case, and some others, simply by deferring table computation until the CC has been built. (Assuming that columns don't depend on datom table selection, anyway.)

Wait until we know types and attributes, and then go. That's equivalent to processing the numeric constraint first.

On the other hand, that's a bit of a hack, and perhaps we should be doing either a multi-pass CC reduction or multi-pass constraint-based algebrizing (basically #373). Walk all the patterns to deduce types and fixed values, then walk them again to produce the ordered constraint list.

I'm not super concerned about this, because real-world queries look like this:

```
[:find ?page ?visit :where
 [?page :visit/timestamp ?timestamp]
 [(> ?timestamp 123456)]]
```

and the attribute is enough to pick a table.

rnewman commented:
I also simplified the type narrowing code:

```
-        if let Some(existing) = self.known_types.get(&variable).cloned() {
-            // If so, the types must match.
+        if let Some(existing) = self.known_types.insert(variable.clone(), unit_type_set(this_type)) {
+            // There was an existing mapping. Does this type match?
             if !existing.contains(&this_type) {
                 self.mark_known_empty(EmptyBecause::TypeMismatch(variable, existing, this_type));
-            } else {
-                if existing.len() > 1 {
-                    // Narrow.
-                    self.known_types.insert(variable, unit_type_set(this_type));
-                }
             }
-        } else {
-            // If not, record the one we just determined.
-            self.known_types.insert(variable, unit_type_set(this_type));
         }
```

rnewman commented:
5e971f3b229f9cc9710706157f7bc9f66799f3f4
